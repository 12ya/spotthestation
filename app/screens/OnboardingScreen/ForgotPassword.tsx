import React, { useState } from "react"
import { ViewStyle, View, Pressable, PressableProps, TextStyle, Linking, Platform } from "react-native"
import { Button, Icon, Text, TextField } from "../../components"
import { Accessory } from "../../components/Accessory"
import { typography } from "../../theme"
import { colors } from "../../theme/colors"

// const { UIMailLauncher } = NativeModules

export interface ForgotPasswordProps {
  /**
   * A function for closing modal.
   */
  onClose?: PressableProps["onPress"]
}

export function ForgotPassword({ onClose }: ForgotPasswordProps) {
  const [isMailed, setIsMailed] = useState(false)
  const [email, setEmail] = useState('')

  const handleSendMail = () => {
   setIsMailed(true)
  }

  const handleOpenEmailApp = () => {
    if (Platform.OS === 'android') {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
      // UIMailLauncher.launchMailApp() //TODO
      return 
    }
    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    Linking.openURL('message:0')
  }

  return (
    <View style={$modalBodyContainer}>
      <Pressable
        style={$close}
        onPress={onClose}
      >
        <Icon icon="x" color={colors.palette.neutral450} />
      </Pressable>
      <View style={$generalIcon}>
        {isMailed ? <Icon icon="mailAccepted" /> : <Icon icon="changePassword" />}
      </View>
      <View style={$contentContainer}>
        <Text tx={`onboarding.forgotPassword.${isMailed ? "titleMailed" : "title"}`} style={$title} />
        <View style={$subtitleContainer}>
          <Text tx={`onboarding.forgotPassword.${isMailed ? "subtitleMailed" : "subtitle"}`} style={$subtitle} />
          {isMailed && <Text text={email} style={$email} />}
        </View>
        {!isMailed && (
          <TextField
            value={email}
            autoCapitalize="none"
            autoComplete="email"
            autoCorrect={false}
            keyboardType="email-address"
            placeholderTx="onboarding.login.placeholder.email"
            onChange={({ nativeEvent }) => setEmail(nativeEvent.text)}
            renderLeftAccessory={({ style }) => <Accessory style={style} icon="mail" />}
            onSubmitEditing={handleSendMail}
            inputWrapperStyle={$textField}
          />
        )}
        <Button
          tx={`onboarding.forgotPassword.${isMailed ? "openEmailAppButton" : "resetButton"}`}
          disabled={!email}
          pressedStyle={$button}
          textStyle={$buttonText}
          style={!email ? [$button, $disabled] : $button}
          onPressIn={isMailed ? handleOpenEmailApp : handleSendMail}
        />
        <View style={$questionContainer}>
          <Text tx={`onboarding.forgotPassword.${isMailed ? "doNotReceiveQuestion" : "rememberQuestion"}`} style={$question} />
          <Pressable onPress={isMailed ? handleSendMail : onClose}>
            <Text tx={`onboarding.forgotPassword.${isMailed ? "resend" : "loginLink"}`} style={$link} />
          </Pressable>
        </View>
      </View>
    </View>
  )
}

const $modalBodyContainer: ViewStyle = {
  backgroundColor: colors.palette.neutral350,
  borderTopLeftRadius: 18,
  borderTopRightRadius: 18,
}

const $contentContainer: ViewStyle = {
  width: "100%",
  paddingHorizontal: 36,
  paddingBottom: 52
}

const $close: ViewStyle = {
  position: "absolute",
  top: 0,
  right: 0,
  padding: 18
}

const $generalIcon: ViewStyle = {
  marginVertical: 36,
  width: "100%",
  alignItems: "center"
}

const $textField: ViewStyle = {
  backgroundColor: colors.palette.neutral550
}

const $questionContainer: TextStyle = {
  flexDirection: "row",
  justifyContent: "center",
}

const $question: TextStyle = {
  color: colors.palette.neutral450,
  fontSize: 18,
  fontFamily: typography.primary.normal,
  lineHeight: 21,
  paddingRight: 3
}

const $title: TextStyle = {
  textAlign: "center",
  color: colors.palette.neutral250,
  fontSize: 36,
  fontFamily: typography.primary.normal,
  lineHeight: 44,
  paddingBottom: 24
}

const $subtitle: TextStyle = {
  textAlign: "center",
  color: colors.palette.neutral450,
  fontSize: 18,
  fontFamily: typography.primary.normal,
  lineHeight: 22
}

const $link: TextStyle = {
  color: colors.palette.neutral250,
  fontSize: 18,
  fontFamily: typography.primary.medium,
  lineHeight: 21,
}

const $email: TextStyle = {
  color: colors.palette.neutral250,
  fontSize: 18,
  fontFamily: typography.primary.medium,
  lineHeight: 22,
  textAlign: "center"
}

const $button: ViewStyle = {
  width: "100%",
  height: 64,
  backgroundColor: colors.palette.buttonBlue,
  borderRadius: 28,
  borderWidth: 0,
  marginTop: 24,
  marginBottom: 24
}

const $buttonText: TextStyle = {
  color: colors.palette.neutral100,
  fontSize: 18,
  fontFamily: typography.primary.medium,
  lineHeight: 22
}

const $disabled: ViewStyle = {
  opacity: .5
}

const $subtitleContainer: ViewStyle = {
  paddingBottom: 36
}