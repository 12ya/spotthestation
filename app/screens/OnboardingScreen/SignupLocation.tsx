import React, { useCallback, useEffect, useRef, useState } from "react"
import { ActivityIndicator, Alert, PressableProps, TextStyle, View, ViewStyle } from "react-native"
import { GooglePlacesAutocomplete, Point, GooglePlacesAutocompleteRef } from "react-native-google-places-autocomplete"
import * as Location from "expo-location"
import { Icon, Text, Button } from "../../components"
import Config from "../../config"
import { translate } from "../../i18n"
import { colors, spacing, typography } from "../../theme"
import { getCurrentLocation } from "../../utils/geolocation"

export interface LocationType {
  title: string,
  subtitle: string,
  location: Point,
  alert?: boolean
}

enum Statuses {
  start = "start",
  detecting = "detecting",
  result = "result",
}

const defaultLocation: LocationType = { title: "Washington, D.C., United States", subtitle: "", location: { lat: 38.89511, lng: -77.03637 }}

export interface SignupLocationProps {
  /**
   * Data about location
   */
  value: LocationType
  /**
   * Action for a change location data
   */
  onValueChange: (value: LocationType) => void
  /**
   * Done button onPress
   */
  onAction: PressableProps["onPress"]
}

export function SignupLocation({ value, onValueChange, onAction }: SignupLocationProps ) {
  const addressRef = useRef<GooglePlacesAutocompleteRef>()
  const [isFocus, setIsFocus] = useState(false)
  const [status, setStatus] = useState(Statuses.start)
  const [textValue, setTextValue] = useState("")

  useEffect(() => {
    if (value.title) setStatus(Statuses.result)
    else if (status === Statuses.result && !value.title) setStatus(Statuses.start)
  }, [value])

  const handleClear = () => {
    addressRef.current?.clear()
    setTextValue("")
  }

  const handleDetect = async () => {
    setStatus(Statuses.detecting)

    if (await Location.hasServicesEnabledAsync()) {
      const location = await getCurrentLocation(() =>
        Alert.alert(
          translate("onboarding.completeProfile.location.permissionAlertTitle"),
          translate("onboarding.completeProfile.location.permissionAlertBody"),
          [{ text: 'OK', onPress: () => {
            onValueChange(defaultLocation)
            setStatus(Statuses.result)
          } }],
          { cancelable: false }
        )
      )
      if (location) onValueChange(location)
    } else {
      Alert.alert(
        translate("onboarding.completeProfile.location.serviceAlertTitle"),
        translate("onboarding.completeProfile.location.serviceAlertBody"),
        [{ text: 'OK', onPress: () => {
          onValueChange(defaultLocation)
          setStatus(Statuses.result)
        } }],
        { cancelable: false }
      )
    }
  }

  const renderStart = useCallback(() => {
    return (
      <>
        <Text
          accessible
          accessibilityLabel="location subtitle"
          accessibilityHint="location subtitle"
          accessibilityRole="text"
          tx="onboarding.completeProfile.location.subtitle" 
          style={$subtitle} 
        />
        <Button
          accessible
          accessibilityLabel="detect button"
          accessibilityHint="detecting current location"
          tx="onboarding.completeProfile.location.detectButton"
          pressedStyle={$button}
          style={$button}
          textStyle={$buttonText}
          onPress={handleDetect}
          renderLeftAccessory={({ style }) => (
            <Icon
              icon="currentLocation"
              size={24}
              style={style}
            />
          )}
        />
        <Text
          accessible
          accessibilityLabel="or"
          accessibilityHint="or"
          accessibilityRole="text"
          tx="onboarding.completeProfile.location.orLabel" 
          style={$orLabel}
        />
        <GooglePlacesAutocomplete
          ref={addressRef}
          placeholder={translate("onboarding.completeProfile.location.selectLocation")}
          query={{
            key: Config.GOOGLE_API_TOKEN,
            language: 'en',
          }}
          onPress={(data, details = null) => onValueChange({ title: details.name, subtitle: details.formatted_address, location: details?.geometry?.location })}
          onFail={(error) => console.error(error)}
          enablePoweredByContainer={false}
          isRowScrollable={false}
          fetchDetails={true}
          keepResultsAfterBlur={true}
          styles={{
            textInputContainer: isFocus ? [$locations, $active] : $locations,
            textInput: {
              ...$dropdownText,
              ...$dropdownSelected,
            },
            listView: $locationsListContainer,
            row: $locationsRow,
            separator: { height: 0 },
            container: { flex: 0, }
          }}
          textInputProps={{
            placeholderTextColor: colors.palette.neutral450,
            onFocus: () => setIsFocus(true),
            onBlur: () => setIsFocus(false),
            onChangeText: (text) => setTextValue(text)
          }}
          renderRow={({ description }) => {
            return <Text text={description} style={$locationsRowText} ellipsizeMode='tail' numberOfLines={1} />
          }}
          renderLeftButton={() => (
            <Icon
              icon="search"
              size={28}
              color={colors.palette.neutral450}
              containerStyle={$dropdownLeftAccessory}
            />
          )}
          renderRightButton={() => isFocus && textValue && (
            <Icon
              icon="xCircle"
              size={28}
              color={colors.palette.neutral450}
              containerStyle={$dropdownRightAccessory}
              onPress={handleClear}
            />
          )}
        />
      </>
    )
  }, [isFocus, textValue])

  const renderDetecting = useCallback(() => {
    return (
      <>
        <View style={$rowContainer}>
          <ActivityIndicator
            accessible
            accessibilityLabel="loader"
            accessibilityHint="Loading"
            accessibilityRole="progressbar"
            animating={status === Statuses.detecting}
            size="small"
            color={colors.palette.neutral450}
            style={$loader}
          />
          <Text 
            accessible
            accessibilityLabel="detecting"
            accessibilityHint="detecting current location"
            accessibilityRole="text"
            tx="onboarding.completeProfile.location.detecting" 
            style={$result}
          />
        </View>
      </>
    )
  }, [status])

  const renderResult = useCallback(() => {
    return (
      <>
        <View style={$rowContainer}>
          <Icon
            icon="pin"
            size={28}
            color={colors.palette.neutral450}
          />
          <Text
            accessible
            accessibilityLabel="address"
            accessibilityHint="display current address"
            accessibilityRole="text"
            text={value.title} 
            style={$result} 
            ellipsizeMode='tail' 
            numberOfLines={1} 
          />
          <Icon
            icon="trash"
            size={28}
            color={colors.palette.neutral450}
            onPress={() => onValueChange({ title: "", subtitle: "", location: { lat: null, lng: null }})}
          />
        </View>
        <Button
          accessible
          accessibilityLabel="Done button"
          accessibilityHint="navigate to main screen"
          tx="onboarding.completeProfile.location.doneButton"
          pressedStyle={$button}
          style={$button}
          textStyle={$buttonText}
          onPress={onAction}
        />
      </>
    )
  }, [value])

  const renderBody = useCallback(() => {
    if (status === Statuses.start) return renderStart()
    if (status === Statuses.detecting) return renderDetecting()
    if (status === Statuses.result) return renderResult()
    return ''
  }, [status, renderStart, renderDetecting, renderResult])

  return (
    <>
      <Text
        accessible
        accessibilityLabel="location title"
        accessibilityHint="location title"
        accessibilityRole="text"
        tx="onboarding.completeProfile.location.tittle"
        style={$title}
      />
      {renderBody()}
    </>
  )
}

const $locations: ViewStyle = {
  borderWidth: 1.5,
  borderColor: "transparent",
  borderRadius: 28,
  height: 56,
  backgroundColor: colors.palette.neutral350,
  overflow: "hidden",
}

const $active: ViewStyle = {
  borderWidth: 1.5,
  borderColor: colors.palette.buttonBlue,
  backgroundColor: colors.palette.overlayBlue,
}

const $locationsListContainer: ViewStyle = {
  borderRadius: 12,
  backgroundColor: colors.palette.neutral550,
  overflow: "hidden",
  width: "85%",
  alignSelf: "center",
  marginTop: 3
}

const $dropdownLeftAccessory: ViewStyle = {
  marginStart: spacing.large,
  height: 56,
  justifyContent: "center",
  alignItems: "center",
}

const $dropdownRightAccessory: ViewStyle = {
  marginEnd: spacing.large,
  height: 56,
  justifyContent: "center",
  alignItems: "center",
}

const $dropdownSelected: TextStyle = {
  color: colors.palette.neutral250,
}

const $locationsRow: TextStyle = {
  backgroundColor: 'transparent',
  paddingHorizontal: spacing.large
}

const $locationsRowText: TextStyle = {
  fontSize: 18,
  lineHeight: 22,
  color: colors.palette.neutral250,
  
}

const $dropdownText: TextStyle = {
  flex: 1,
  height: 56,
  alignSelf: "stretch",
  fontFamily: typography.primary.normal,
  fontSize: 18,
  lineHeight: 22,
  paddingHorizontal: 0,
  marginHorizontal: spacing.small,
  textAlignVertical: "center",
  borderRadius: 0,
  backgroundColor: "transparent"
}

const $title: TextStyle = {
  color: colors.palette.neutral250,
  fontSize: 36,
  fontFamily: typography.primary.normal,
  lineHeight: 44,
  paddingBottom: 24,
}

const $subtitle: TextStyle = {
  color: colors.palette.neutral450,
  fontSize: 18,
  fontFamily: typography.primary.normal,
  lineHeight: 22,
  paddingBottom: 36,
}

const $orLabel: TextStyle = {
  color: colors.palette.neutral450,
  fontSize: 16,
  fontFamily: typography.primary.normal,
  lineHeight: 19,
  textAlign: "center",
  marginVertical: 12
}

const $button: ViewStyle = {
  width: "100%",
  height: 56,
  backgroundColor: colors.palette.buttonBlue,
  borderRadius: 28,
  borderWidth: 0
}

const $buttonText: TextStyle = {
  color: colors.palette.neutral100,
  fontSize: 18,
  fontFamily: typography.primary.medium,
  lineHeight: 21,
}

const $rowContainer: ViewStyle = {
  width: "100%",
  flexDirection: "row",
  justifyContent: "space-between"
}

const $result: TextStyle = {
  flex: 1,
  paddingHorizontal: 18,
  color: colors.palette.neutral250,
  fontSize: 18,
  fontFamily: typography.primary.normal,
  lineHeight: 22,
  paddingBottom: 56,
}

const $loader: ViewStyle = {
  justifyContent: "flex-start"
}
