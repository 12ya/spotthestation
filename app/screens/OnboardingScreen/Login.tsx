import { Link } from "@react-navigation/native"
import React, { useCallback, useMemo, useRef, useState } from "react"
import { ViewStyle, TextStyle, View, TextInput } from "react-native"
import Modal from "react-native-modal"
import { Button, Icon, Screen, Text, TextField, TextFieldAccessoryProps } from "../../components"
import { Accessory } from "../../components/Accessory"
import { typography } from "../../theme"
import { colors } from "../../theme/colors"
import { useSafeAreaInsetsStyle } from "../../utils/useSafeAreaInsetsStyle"
import { IconLinkButton } from "./components/IconLinkButton"
import { NasaLogo } from "./components/NasaLogo"
import { ForgotPassword } from "./ForgotPassword"

export function Login() {
  const $topInset = useSafeAreaInsetsStyle(["top", "bottom"], "padding")

  const passwordInput = useRef<TextInput>()

  const [email, setEmail] = useState('')
  const [password, setPassword] = useState('')
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const [loginError, setLoginError] = useState('')
  const [isPasswordHidden, setIsPasswordHidden] = useState(false)
  const [isForgotPassword, setIsForgotPassword] = useState(false)

  const isCanLogin = useMemo(() => email && password, [email, password])

  const handleLogin = useCallback(() => {
    if (!isCanLogin) return
    console.log('TODO')
  }, [isCanLogin])

  const PasswordRightAccessory = useMemo(
    () =>
      function PasswordRightAccessory(props: TextFieldAccessoryProps) {
        return (
          <>
            {Boolean(loginError) && <Icon
              icon="delete"
              size={28}
              color={colors.palette.neutral450}
              containerStyle={props.style}
              onPress={() => setPassword('')}
            />}
            <Icon
              icon={isPasswordHidden ? "view" : "hidden"}
              size={28}
              color={colors.palette.neutral450}
              containerStyle={props.style}
              onPress={() => setIsPasswordHidden(!isPasswordHidden)}
            />
          </>
        )
      },
    [isPasswordHidden, loginError],
  )

  return (
    <Screen
      preset="fixed"
      contentContainerStyle={$container}
      statusBarStyle="light"
      style={[$topInset, {backgroundColor: colors.palette.neutral900}]}
    >
      <NasaLogo />
      <Text tx={`onboarding.login.${loginError ? 'loginError' : 'title'}`} style={$title} />
      <View style={$bodyContainer}>
        <View style={$inputsContainer}>
          <TextField
            value={email}
            autoCapitalize="none"
            autoComplete="email"
            autoCorrect={false}
            keyboardType="email-address"
            placeholderTx="onboarding.login.placeholder.email"
            onChange={({ nativeEvent }) => setEmail(nativeEvent.text)}
            inputWrapperStyle={loginError ? $error : {}}
            LeftAccessory={(props) => (
              <Accessory 
                props={props} 
                icon="mail"
                color={loginError ? colors.palette.nasaRed : colors.palette.neutral450} 
              />
            )}
            RightAccessory={(props) => loginError && (
              <Accessory 
                props={props} 
                icon="delete"
                onPress={() => setEmail('')}
              />
            )}
            onSubmitEditing={() => passwordInput.current?.focus()}
          />
          <TextField
            ref={passwordInput}
            value={password}
            autoCapitalize="none"
            autoComplete="password"
            autoCorrect={false}
            secureTextEntry={isPasswordHidden}
            placeholderTx="onboarding.login.placeholder.password"
            onChange={({ nativeEvent }) => setPassword(nativeEvent.text)}
            inputWrapperStyle={loginError ? $error : {}}
            RightAccessory={PasswordRightAccessory}
            LeftAccessory={(props) => (
              <Accessory 
                props={props} 
                icon="lock"
                color={loginError ? colors.palette.nasaRed : colors.palette.neutral450} 
              />
            )}
            onSubmitEditing={handleLogin}
          />
        </View>
        <View style={$loginButtonContainer}>
          <Button tx={"onboarding.login.forgotPassword"}
            pressedStyle={$forgotButton}
            textStyle={$forgot}
            style={$forgotButton}
            onPressIn={() => setIsForgotPassword(true)}
          />
          <Button tx="onboarding.login.loginButton"
            disabled={!isCanLogin}
            pressedStyle={$loginButton}
            textStyle={$loginButtonText}
            style={!isCanLogin ? [$loginButton, $disabled] : $loginButton}
            onPressIn={handleLogin}
          />
        </View>
        <View style={$loginWithContainer}>
          <Text tx={"onboarding.login.loginWith"} style={$loginWith} />
          <View style={$loginWithButtonsContainer}>
            <IconLinkButton icon="brandGoogle" buttonStyle={$buttonSize} />
            <IconLinkButton icon="brandFacebook" buttonStyle={$buttonSize} />
            <IconLinkButton icon="brandTwitter" buttonStyle={$buttonSize} />
          </View>
        </View>
        <View style={$questionContainer}>
          <Text tx={"onboarding.login.haveAccountQuestion"} style={$question} />
          <Link to={{ screen: "Signup" }}>
            <Text tx={"onboarding.login.signUpLink"} style={$link} />
          </Link>
        </View>
      </View>
      <Modal
        isVisible={isForgotPassword}
        onBackdropPress={() => setIsForgotPassword(!isForgotPassword)}
        onSwipeComplete={() => setIsForgotPassword(!isForgotPassword)}
        animationIn="slideInUp"
        animationOut="slideOutDown"
        swipeDirection="down"
        useNativeDriver
        useNativeDriverForBackdrop
        hideModalContentWhileAnimating
        propagateSwipe
        backdropOpacity={0.65}
        style={$modal}
      >
        <ForgotPassword onClose={() => setIsForgotPassword(!isForgotPassword)} />
      </Modal>
    </Screen>
  )
}

const $container: ViewStyle = {
  flex: 1,
  backgroundColor: colors.palette.neutral900,
}

const $modal: ViewStyle = {
  flex: 1,
  justifyContent: 'flex-end',
  left: 0,
  margin: 0
}

const $bodyContainer: ViewStyle = {
  flex: 1,
  justifyContent: "space-between",
  paddingTop: 36,
  paddingBottom: 12,
}

const $title: TextStyle = {
  color: colors.palette.neutral250,
  fontSize: 36,
  fontFamily: typography.primary.normal,
  lineHeight: 42,
  paddingLeft: 36,
  paddingTop: 36
}

const $loginButtonContainer: ViewStyle = {
  flexDirection: "row",
  justifyContent: "space-between",
  alignItems: "center",
  paddingHorizontal: 36
}

const $forgot: TextStyle = {
  color: colors.palette.neutral100,
  fontSize: 18,
  fontFamily: typography.primary.normal,
  lineHeight: 21,
}

const $forgotButton: ViewStyle = {
  backgroundColor: "transparent",
  borderWidth: 0
}

const $loginButton: ViewStyle = {
  width: 140,
  height: 56,
  backgroundColor: colors.palette.buttonBlue,
  borderRadius: 28,
  borderWidth: 0
}

const $loginButtonText: TextStyle = {
  color: colors.palette.neutral100,
  fontSize: 18,
  fontFamily: typography.primary.medium,
  lineHeight: 21,
}

const $loginWithContainer: ViewStyle = {
  width: "60%",
  alignSelf: "center",
  alignItems: "center",
}

const $loginWithButtonsContainer: ViewStyle = {
  width: "100%",
  flexDirection: "row",
  justifyContent: "space-between",
  paddingTop: 24
}

const $loginWith: TextStyle = {
  color: colors.palette.neutral450,
  fontSize: 15,
  fontFamily: typography.primary.normal,
  lineHeight: 19,
}

const $buttonSize: TextStyle = {
  width: 64,
  height: 64
}

const $questionContainer: TextStyle = {
  flexDirection: "row",
  justifyContent: "center",
}

const $question: TextStyle = {
  color: colors.palette.neutral450,
  fontSize: 18,
  fontFamily: typography.primary.normal,
  lineHeight: 21,
  paddingRight: 3
}

const $link: TextStyle = {
  color: colors.palette.neutral250,
  fontSize: 18,
  fontFamily: typography.primary.medium,
  lineHeight: 21,
}

const $disabled: ViewStyle = {
  opacity: .5
}

const $error: ViewStyle = {
  backgroundColor: colors.palette.inputError
}

const $inputsContainer: ViewStyle = {
  paddingHorizontal: 36,
  justifyContent: "space-between",
  height: 150
}
