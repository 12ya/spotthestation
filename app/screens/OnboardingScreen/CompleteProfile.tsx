import { useNavigation } from "@react-navigation/native"
import React, { useCallback, useState } from "react"
import { TextStyle, View, ViewStyle } from "react-native"
import { Screen, Text } from "../../components"
import { typography } from "../../theme"
import { colors } from "../../theme/colors"
import { useSafeAreaInsetsStyle } from "../../utils/useSafeAreaInsetsStyle"
import { IconLinkButton } from "./components/IconLinkButton"
import { LocationType, SignupLocation } from "./SignupLocation"
import { SignupNotificationSettings } from "./SignupNotificationSettings"
import * as storage from "../../utils/storage"

export function CompleteProfile() {
  const navigation = useNavigation()

  const $topInset = useSafeAreaInsetsStyle(["top", "bottom"], "padding")

  const [step, setStep] = useState(1)
  const [notifications, setNotifications] = useState(false)
  const [location, setLocation] = useState({
    title: "",
    subtitle: "",
    location: {
      lat: null,
      lng: null
    }
  })

  const handleBack = useCallback(() => {
    if (step === 1) navigation.goBack()
    else setStep(step - 1)
  }, [step])

  const handleNext = useCallback(() => setStep(step + 1), [step])

  const handleDone = () => navigation.navigate("Main" as never)

  const handleLocationChange = async (location: LocationType) => {
    setLocation(location)
    await storage.save('currentLocation', location)
  }

  const renderBody = useCallback(() => {
    switch(step) {
      case 2: return <SignupLocation value={location} onValueChange={handleLocationChange} onAction={handleDone} />
      default: return <SignupNotificationSettings value={notifications} onValueChange={setNotifications} onAction={handleNext} />
    }
  }, [step, notifications, location])

  return (
    <Screen preset="scroll" contentContainerStyle={$container} style={[$topInset, {backgroundColor: colors.palette.neutral900}]} statusBarStyle="light">
      <IconLinkButton 
        icon="back"
        accessible
        accessibilityLabel="Back button"
        accessibilityHint="Navigates to the previous screen"
        onPress={handleBack} 
        buttonStyle={$back} 
      />
      <View style={$contentContainer}>
        <Text
          accessible
          accessibilityLabel="Step counter"
          accessibilityHint="Display current step"
          accessibilityRole="text"
          text={`${step} of 2`} 
          style={$step}
        />
        {renderBody()}
      </View>
    </Screen>
  )
}

const $container: ViewStyle = {
  flex: 1,
  backgroundColor: colors.palette.neutral900,
  paddingTop: 24
}

const $contentContainer: ViewStyle = {
  flex: 1,
  paddingHorizontal: 36,
  paddingBottom: 52
}

const $back: ViewStyle = {
  marginLeft: 18
}

const $step: TextStyle = {
  color: colors.palette.neutral450,
  fontSize: 18,
  fontFamily: typography.primary.normal,
  lineHeight: 22,
  paddingBottom: 18,
  paddingTop: 36
}
