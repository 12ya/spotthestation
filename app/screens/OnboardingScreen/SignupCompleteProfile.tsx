import { useNavigation } from "@react-navigation/native"
import React, { useCallback, useState } from "react"
import { TextStyle, View, ViewStyle } from "react-native"
import { Screen, Text } from "../../components"
import { typography } from "../../theme"
import { colors } from "../../theme/colors"
import { useSafeAreaInsetsStyle } from "../../utils/useSafeAreaInsetsStyle"
import { IconLinkButton } from "./components/IconLinkButton"
import { SignupLocation } from "./SignupLocation"
import { SignupNotificationSettings } from "./SignupNotificationSettings"
import { FieldTypes, SignupProfile } from "./SignupProfile"

export function SignupCompleteProfile() {
  const navigation = useNavigation()

  const $topInset = useSafeAreaInsetsStyle(["top", "bottom"], "padding")

  const [step, setStep] = useState(1)
  const [profileData, setProfileData] = useState({
    firstName: "",
    lastName: "",
    country: "",
    state: "",
    city: ""
  })
  const [notifications, setNotifications] = useState(false)
  const [location, setLocation] = useState({
    title: "",
    location: {
      lat: null,
      lng: null
    }
  })

  const handleBack = useCallback(() => {
    if (step === 1) navigation.goBack()
    else setStep(step - 1)
  }, [step])

  const handleNext = useCallback(() => setStep(step + 1), [step])

  const handleDone = () => navigation.navigate("Main" as never)

  const handleChangeProfile = useCallback((value: string, field: FieldTypes) => {
    setProfileData({...profileData, [field]: value })
  }, [profileData])

  const renderBody = useCallback(() => {
    switch(step) {
      case 2: return <SignupNotificationSettings value={notifications} onValueChange={setNotifications} onAction={handleNext} />
      case 3: return <SignupLocation value={location} onValueChange={setLocation} onAction={handleDone} />
      default: return <SignupProfile value={profileData} onValueChange={handleChangeProfile} onAction={handleNext} />
    }
  }, [step, profileData, notifications, location])

  return (
    <Screen preset="fixed" contentContainerStyle={$container} style={[$topInset, {backgroundColor: colors.palette.neutral900}]} statusBarStyle="light">
      <IconLinkButton icon="back" onPress={handleBack} buttonStyle={$back} />
      <View style={$contentContainer}>
        <Text text={`${step} of 3`} style={$step} />
        {renderBody()}
      </View>
    </Screen>
  )
}

const $container: ViewStyle = {
  flex: 1,
  backgroundColor: colors.palette.neutral900,
  paddingTop: 24
}

const $contentContainer: ViewStyle = {
  flex: 1,
  paddingHorizontal: 36,
  paddingBottom: 52
}

const $back: ViewStyle = {
  marginLeft: 18
}

const $step: TextStyle = {
  color: colors.palette.neutral450,
  fontSize: 18,
  fontFamily: typography.primary.normal,
  lineHeight: 22,
  paddingBottom: 18,
  paddingTop: 36
}
