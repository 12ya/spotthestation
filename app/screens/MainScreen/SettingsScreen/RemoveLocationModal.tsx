import React from "react"
import { ViewStyle, View, PressableProps, TextStyle } from "react-native"
import { Icon, Text, Button } from "../../../components"
import { colors, typography } from "../../../theme"
import { LocationType } from "../../OnboardingScreen/SignupLocation"

export interface RemoveLocationModalProps {
  onClose?: PressableProps["onPress"]
  onRemove?: PressableProps["onPress"]
  location: LocationType
}

export function RemoveLocationModal({ onClose, onRemove, location }: RemoveLocationModalProps) {
  return (
    <View style={$modalBodyContainer}>
      <Icon icon="x" 
        accessible
        accessibilityLabel="x button"
        accessibilityHint="close modal"
        accessibilityRole="button"
        color={colors.palette.neutral450} 
        onPress={onClose} 
        containerStyle={$close} 
        size={36}
      />
      <View style={$generalIcon}>
        <Icon icon="removeLocation" />
      </View>
      <View style={$contentContainer}>
        <Text tx="settings.locationSettingsData.removeLocation.question" style={$title} />
        <Text text={location?.title} style={$locationTitle} />
        <Text text={location?.subtitle} style={$locationAddress} />
        
        <View style={$buttonsContainer}>
          <Button
            tx="settings.locationSettingsData.removeLocation.cancelButton"
            pressedStyle={$button}
            textStyle={$buttonText}
            style={$button}
            onPressIn={onClose}
          />
          <Button
            tx="settings.locationSettingsData.removeLocation.removeButton"
            pressedStyle={[$button, { backgroundColor: colors.palette.nasaRed }]}
            textStyle={$buttonText}
            style={[$button, { backgroundColor: colors.palette.nasaRed }]}
            onPressIn={onRemove}
          />
        </View>
      </View>
    </View>
  )
}

const $modalBodyContainer: ViewStyle = {
  backgroundColor: colors.palette.neutral350,
  borderTopLeftRadius: 18,
  borderTopRightRadius: 18,
}

const $buttonsContainer: ViewStyle = {
  flexDirection: 'row',
  alignItems: "center",
  justifyContent: "space-between",
}

const $contentContainer: ViewStyle = {
  width: "100%",
  paddingHorizontal: 36,
  paddingBottom: 52
}

const $close: ViewStyle = {
  position: "absolute",
  top: 0,
  right: 0,
  padding: 18,
  zIndex: 5
}

const $generalIcon: ViewStyle = {
  marginVertical: 36,
  width: "100%",
  alignItems: "center"
}

const $locationTitle: TextStyle = {
  textAlign: "center",
  color: colors.palette.neutral250,
  fontSize: 18,
  fontFamily: typography.primary.medium,
  lineHeight: 22,
  paddingBottom: 5
}

const $title: TextStyle = {
  textAlign: "center",
  color: colors.palette.neutral450,
  fontSize: 18,
  fontFamily: typography.primary.normal,
  lineHeight: 22,
  paddingBottom: 24
}

const $locationAddress: TextStyle = {
  textAlign: "center",
  color: colors.palette.neutral450,
  fontSize: 16,
  fontFamily: typography.primary.normal,
  lineHeight: 19,
  paddingBottom: 36
}

const $button: ViewStyle = {
  width: '40%',
  height: 64,
  backgroundColor: colors.palette.neutral550,
  borderRadius: 28,
  borderWidth: 0,
  marginTop: 24,
  marginBottom: 24
}

const $buttonText: TextStyle = {
  color: colors.palette.neutral100,
  fontSize: 18,
  fontFamily: typography.primary.medium,
  lineHeight: 22
}
