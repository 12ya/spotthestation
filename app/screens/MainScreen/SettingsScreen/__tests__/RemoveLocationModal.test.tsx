/* eslint-disable react/display-name */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import React from "react"
import renderer from 'react-test-renderer'
import { RemoveLocationModal } from '../RemoveLocationModal'
import { LocationType } from "../../../OnboardingScreen/SignupLocation"

it('renders correctly', () => {
  const tree = renderer
    .create(<RemoveLocationModal location={{} as LocationType} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
