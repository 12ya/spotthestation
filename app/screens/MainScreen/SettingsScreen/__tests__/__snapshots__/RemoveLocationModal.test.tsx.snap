// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#1E2022",
      "borderTopLeftRadius": 26,
      "borderTopRightRadius": 26,
    }
  }
>
  <div />
  <View
    style={
      Object {
        "alignItems": "center",
        "marginVertical": 52,
        "width": "100%",
      }
    }
  >
    <div />
  </View>
  <View
    style={
      Object {
        "paddingBottom": 75,
        "paddingHorizontal": 52,
        "width": "100%",
      }
    }
  >
    <div />
    <div />
    <div />
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <div />
      <div />
    </View>
  </View>
</View>
`;
